### PR TITLE
Remove Quaterniond and use Matrix3d

### LINKED ISSUE #1214: Eigen::Quaterniond memory issues
`Quaterniond` is https://eigen.tuxfamily.org/dox/group__TopicFixedSizeVectorizable.html (unlike `Vector3d`, which is not). Address-sanitizer finds issues (double free or corruption) in `VariableTest::rotate`. It is no not entirely clear if this is due to an alignment issues in memory allocation from `element_array`, or something related to passing `Quaterniond` by value, which is also forbidden.

Possible solutions:
- Use 3x3 matrix instead of quaterion.
- Use aligned allocation. This is not as simple as may be initially thought: Need to allocate aligned memory (which is easier in C++17), but also use, e.g., placement-new to ensure constructors are called, and ensure destructors are called manually before freeing the memory (manually). Overall, quite a lot of things that one could get wrong in the implementation.
- Do not use `Eigen`, either implement our own quaternion, or matrices and vectors. The two sub-options here would be to do this using a custom quaternion and vector class, or to use internal extra dimensions in our arrays. The latter would be a bigger change across the board, but allow for variable-sized vector suppoert.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #1214 by avoiding fixed-size-vectorizable type.

The memory overhead in practice should be negligible for our applications. I notice that there are no detailed tests for the rotation handling in the Mantid converter. Docs appear to build correctly, but maybe there is something there reviewer should test by hand?

### CODE REVIEW

File: python/tests/compat/test_mantid.py [hunk 1]
```diff
@@ -528,6 +515,15 @@ def test_validate_units_throws(self):
                 mantidcompat.validate_dim_and_get_mantid_string(i)
 
 
+def test_to_quat_from_vectors():
```
Comment (OwenArnold) [inline python/tests/compat/test_mantid.py:518]: rename `test_to_rot_from_vectors`

Comment (SimonHeybrock) [reply python/tests/compat/test_mantid.py:518]:
```suggestion
def test_to_rot_from_vectors():
```

**Comment by SimonHeybrock** [comment]
> Suggestion - Mantid's Quat can yield Euler angles and I'm wondering if using quaternion at all is necessary. Likely easier for users to have the Euler alternative. Also potentially easier for computing angles between two vector as required for spherical coordinate conversion.

All of this has to be rewritten entirely. The manual conversion of rotation matrices is unusably slow, which is why we disabled it (must use `advanced_geometry=True`). I was just now in a call with instrument scientists and they basically stated they do not care about seeing the actual pixel shapes (making rotations unnecessary, as long as we do not not shapes for anything else).

**Comment by SimonHeybrock** [comment]
@OwenArnold I think your comments has been addressed.

**Pull request merged**